sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -1395,7 +1395,7 @@ public DataSource apply(Void input) {
 
     static synchronized DataSource buildDataSource(Void input) {
       if (source == null) {
-        DataSource basicSource = dataSourceProviderFn.apply(input);
+        DataSource basicSource = dataSourceProviderFn.apply(null);
```

Comment:
TBH this is a bit mysterious if it fixes anything, since `null` is the only constructible value of type `Void`.

Reply:
Agreed. I'm not convinced this is a solution. Going to do some testing.

Reply:
Looks like this change has no impact. I think `dataSourceProviderFn` is null when `buildDataSource(Void input)` is called via apply. There is an explicit test for that: https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java#L184-L190

Reply:
The issue is that PoolableDataSourceProvider stores dataSourceProviderFn as a static which means that it will never be serialized and once it is deserialized it is lost.